// @mailchain/internal/node/sending/mail/convertSendMailParams.js
'use strict';

var crypto = require('@mailchain/crypto');
var encoding = require('@mailchain/encoding');

/**
 * Temporary adapter migration function to convert from SDK mail params to internal params.
 *
 * @deprecated This will be removed in favor of removing the internal params and accepting these params.
 */
function toMailData(params) {
    const mailDomain = params.from.split('@')[1];
    return {
        id: `${encoding.encodeBase64(crypto.secureRandom(32))}@${mailDomain}`,
        date: new Date(),
        subject: params.subject,
        from: createMailAddress(params.from),
        replyTo: params.replyTo ? createMailAddress(params.replyTo) : undefined,
        recipients: params.to?.map((to) => createMailAddress(to)) ?? [],
        carbonCopyRecipients: params.cc?.map((cc) => createMailAddress(cc)) ?? [],
        blindCarbonCopyRecipients: params.bcc?.map((bcc) => createMailAddress(bcc)) ?? [],
        message: params.content.html?.toString() ?? '',
        plainTextMessage: params.content.text?.toString() ?? '',
    };
}
/**
 * Temporary adapter migration function to convert from internal MailData to SDK mail params.
 *
 * @deprecated This will be removed in favor of removing the internal params and accepting these params.
 */
function createMailAddress(address) {
    const name = address.split('@')[0];
    return { name, address };
}

exports.createMailAddress = createMailAddress;
exports.toMailData = toMailData;
