// @mailchain/addressing/node/protocols/consts.js
'use strict';

// WARNING: CAREFUL WHEN CHANGING THE VALUES OF THIS CONSTANTS, THEY MIGHT BE ALREADY LINKED TO KEY PATHS
/**
 * Algorand protocol name.
 */
const ALGORAND = 'algorand';
/**
 * Ethereum protocol name.
 */
const ETHEREUM = 'ethereum';
/**
 * Substrate protocol name.
 */
const SUBSTRATE = 'substrate';
const NEAR = 'near';
const TEZOS = 'tezos';
const FILECOIN = 'filecoin';
/**
 * Mailchain protocol name.
 */
const MAILCHAIN = 'mailchain';
const ALL_PROTOCOLS = [ALGORAND, ETHEREUM, SUBSTRATE, NEAR, MAILCHAIN, TEZOS, FILECOIN];
const ENABLED_PROTOCOLS = [ETHEREUM, NEAR, TEZOS, FILECOIN];
function isBlockchainProtocolEnabled(protocol) {
    return ENABLED_PROTOCOLS.includes(protocol);
}

exports.ALGORAND = ALGORAND;
exports.ALL_PROTOCOLS = ALL_PROTOCOLS;
exports.ETHEREUM = ETHEREUM;
exports.FILECOIN = FILECOIN;
exports.MAILCHAIN = MAILCHAIN;
exports.NEAR = NEAR;
exports.SUBSTRATE = SUBSTRATE;
exports.TEZOS = TEZOS;
exports.isBlockchainProtocolEnabled = isBlockchainProtocolEnabled;
