// @mailchain/internal/browser/sending/mail/payloads.js
import { KindNaClSecretKey } from '@mailchain/crypto';
import { EncodingTypes } from '@mailchain/encoding';
import { createMimeMessage } from '../../formatters/generate.js';

async function createMailPayloads(senderMessagingKey, resolvedAddresses, mailData) {
    const message = await createMimeMessage(mailData, resolvedAddresses);
    const original = await createMailPayload(senderMessagingKey, message.original);
    const visibleRecipientsPayload = await createMailPayload(senderMessagingKey, message.visibleRecipients);
    const blindRecipients = await Promise.all(message.blindRecipients.map(async ({ recipient, content }) => ({
        recipients: [recipient],
        payload: await createMailPayload(senderMessagingKey, content),
    })));
    return {
        original,
        distributions: [
            {
                recipients: [...mailData.recipients, ...mailData.carbonCopyRecipients],
                payload: visibleRecipientsPayload,
            },
            ...blindRecipients,
        ],
    };
}
async function createMailPayload(senderMessagingKey, contentPayload) {
    const contentBuffer = Buffer.from(contentPayload);
    return {
        Headers: {
            Origin: senderMessagingKey.publicKey,
            ContentSignature: await senderMessagingKey.sign(contentBuffer),
            Created: new Date(),
            ContentLength: contentBuffer.length,
            ContentType: 'message/x.mailchain',
            ContentEncoding: EncodingTypes.Base64,
            ContentEncryption: KindNaClSecretKey,
        },
        Content: contentBuffer,
    };
}

export { createMailPayload, createMailPayloads };
