// @mailchain/internal/browser/sending/mail/prepare.js
import { SenderMessagingKeyIncorrect } from '@mailchain/signatures';
import { PreflightCheckError } from '../errors.js';
import { createMailPayloads } from './payloads.js';
import { MessagingKeys } from '../../messagingKeys/messagingKeys.js';
import { MailSenderVerifier } from '../../transport/mail/sender.js';

class MailPreparer {
    constructor(messagingKeys, mailSenderVerifier) {
        Object.defineProperty(this, "messagingKeys", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: messagingKeys
        });
        Object.defineProperty(this, "mailSenderVerifier", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: mailSenderVerifier
        });
    }
    static create(configuration) {
        return new MailPreparer(MessagingKeys.create(configuration), MailSenderVerifier.create(configuration));
    }
    async prepareMail(params) {
        const { message, senderMessagingKey } = params;
        if (message.subject.length === 0) {
            return { error: new PreflightCheckError('Subject must not be empty.') };
        }
        if (message.plainTextMessage.length === 0) {
            return { error: new PreflightCheckError('Content text must not be empty.') };
        }
        if (message.message.length === 0) {
            return { error: new PreflightCheckError('Content html must not be empty.') };
        }
        const allRecipients = [
            ...message.recipients,
            ...message.blindCarbonCopyRecipients,
            ...message.carbonCopyRecipients,
        ];
        if (allRecipients.length === 0) {
            return { error: new PreflightCheckError('No recipients found.') };
        }
        const isSenderMatching = await this.mailSenderVerifier.verifySenderOwnsFromAddress(message.from, senderMessagingKey.publicKey);
        if (!isSenderMatching) {
            return { error: new SenderMessagingKeyIncorrect() };
        }
        // add at after checking if all recipients are empty
        const allParticipants = [...allRecipients, message.from];
        if (message.replyTo != null) {
            allParticipants.push(message.replyTo);
        }
        const { data: resolvedAddresses, error } = await this.messagingKeys.resolveMany(allParticipants.map((x) => x.address));
        if (error) {
            return { error };
        }
        const messagePayloads = await createMailPayloads(params.senderMessagingKey, resolvedAddresses, message);
        return {
            data: {
                distributions: messagePayloads.distributions,
                message: messagePayloads.original,
                resolvedAddresses,
            },
        };
    }
}

export { MailPreparer };
