// @mailchain/internal/browser/sending/mail/sender.js
import { defaultConfiguration } from '../../configuration.js';
import { toMailData } from './convertSendMailParams.js';
import { MailPreparer } from './prepare.js';
import { MailDistributor } from './distributor.js';

class MailSender {
    constructor(senderMessagingKey, mailPreparer, mailDistributor) {
        Object.defineProperty(this, "senderMessagingKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: senderMessagingKey
        });
        Object.defineProperty(this, "mailPreparer", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: mailPreparer
        });
        Object.defineProperty(this, "mailDistributor", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: mailDistributor
        });
    }
    static fromSenderMessagingKey(senderMessagingKey, configuration = defaultConfiguration) {
        return new MailSender(senderMessagingKey, MailPreparer.create(configuration), MailDistributor.create(configuration, senderMessagingKey));
    }
    /**
     * Send a mail to any blockchain or Mailchain address using the address Messaging Key.
     *
     * @param params {@link SendMailParams} - The parameters for sending a mail.
     * @returns
     */
    async sendMail(params) {
        const { data: preparedMail, error: prepareMailError } = await this.mailPreparer.prepareMail({
            message: toMailData(params),
            senderMessagingKey: this.senderMessagingKey,
        });
        if (prepareMailError) {
            return { error: prepareMailError };
        }
        const { data: distributedMail, error: distributedMailError } = await this.mailDistributor.distributeMail({
            distributions: preparedMail.distributions,
            resolvedAddresses: preparedMail.resolvedAddresses,
        });
        if (distributedMailError) {
            return { error: distributedMailError };
        }
        return {
            data: {
                sentMailDeliveryRequests: distributedMail,
            },
        };
    }
}

export { MailSender };
