// @mailchain/addressing/browser/index.js
export { isSameAddress } from './addressComparison.js';
export { formatAddress } from './addressFormatting.js';
export { addressFromPublicKey } from './addressFromPublicKey.js';
export { isEthereumAddress, isFilecoinAddress, isMailchainAccountAddress, isNearImplicitAccount, isTezosAddress } from './addressPredicates.js';
export { checkAddressForErrors } from './checkAddressForErrors.js';
export { decodeAddressByProtocol, encodeAddressByProtocol, encodingByProtocol } from './encoding.js';
export { BadlyFormattedAddressError, IdentityProviderAddressInvalidError } from './errors.js';
export { EthereumAlice, EthereumBob } from './test.constants.js';
export { parseWalletAddress } from './parseWalletAddress.js';
export { createNameServiceAddress, isNameServiceAddress, isSameNameServiceAddress, parseNameServiceAddress } from './nameServiceAddress.js';
export { createWalletAddress } from './walletAddress.js';
export { ALGORAND, ALL_PROTOCOLS, ETHEREUM, FILECOIN, MAILCHAIN, NEAR, SUBSTRATE, TEZOS, isBlockchainProtocolEnabled } from './protocols/consts.js';
export { ProtocolNotSupportedError } from './protocols/errors.js';
export { guessProtocolsFromAddress } from './protocols/protocols.js';
