// @mailchain/signatures/node/index.js
'use strict';

var errors = require('./errors.js');
var eth_personal = require('./eth_personal.js');
var mailchain_message_confirmation = require('./mailchain_message_confirmation.js');
var mailchain_msgkey = require('./mailchain_msgkey.js');
var mailchain_password_reset = require('./mailchain_password_reset.js');
var mailchain_username = require('./mailchain_username.js');
var raw_ed25519 = require('./raw_ed25519.js');
var verify = require('./verify.js');
var mailer = require('./mailer.js');



exports.AddressIsEmptyError = errors.AddressIsEmptyError;
exports.AddressMustBeProtocolAddressError = errors.AddressMustBeProtocolAddressError;
exports.MessagingKeyVerificationError = errors.MessagingKeyVerificationError;
exports.ProtocolIsEmptyError = errors.ProtocolIsEmptyError;
exports.PublicKeyNotFoundError = errors.PublicKeyNotFoundError;
exports.SenderMessagingKeyIncorrect = errors.SenderMessagingKeyIncorrect;
exports.getMessageHash = eth_personal.getMessageHash;
exports.signEthereumPersonalMessage = eth_personal.signEthereumPersonalMessage;
exports.verifyEthereumPersonalMessage = eth_personal.verifyEthereumPersonalMessage;
exports.mailchainDeliveryConfirmationMessage = mailchain_message_confirmation.mailchainDeliveryConfirmationMessage;
exports.signMailchainDeliveryConfirmation = mailchain_message_confirmation.signMailchainDeliveryConfirmation;
exports.mailchainProvidedMessagingKeyMessage = mailchain_msgkey.mailchainProvidedMessagingKeyMessage;
exports.signMailchainProvidedMessagingKey = mailchain_msgkey.signMailchainProvidedMessagingKey;
exports.verifyMailchainProvidedMessagingKey = mailchain_msgkey.verifyMailchainProvidedMessagingKey;
exports.mailchainPasswordResetMessage = mailchain_password_reset.mailchainPasswordResetMessage;
exports.signMailchainPasswordReset = mailchain_password_reset.signMailchainPasswordReset;
exports.verifyMailchainPasswordReset = mailchain_password_reset.verifyMailchainPasswordReset;
exports.signMailchainUsername = mailchain_username.signMailchainUsername;
exports.verifyMailchainUsername = mailchain_username.verifyMailchainUsername;
exports.signRawEd25519 = raw_ed25519.signRawEd25519;
exports.verifyRawEd25519 = raw_ed25519.verifyRawEd25519;
exports.verify = verify.verify;
exports.createMailerProof = mailer.createMailerProof;
exports.createMailerProofSigningData = mailer.createMailerProofSigningData;
exports.signMailerProofParams = mailer.signMailerProofParams;
exports.verifyMailerProof = mailer.verifyMailerProof;
