// @mailchain/crypto/node/index.js
'use strict';

var keys = require('./keys.js');
var _public$2 = require('./public.js');
var rand = require('./rand.js');
var hd$1 = require('./hd.js');
var errors = require('./errors.js');
var hd = require('./ed25519/hd.js');
var _private = require('./ed25519/private.js');
var _public = require('./ed25519/public.js');
var derivation = require('./ed25519/derivation.js');
var _private$1 = require('./secp256k1/private.js');
var _public$1 = require('./secp256k1/public.js');
var names = require('./multikey/names.js');
var bytes = require('./multikey/bytes.js');
var ids = require('./multikey/ids.js');
var ed25519 = require('./cipher/ecdh/ed25519.js');
var cipher = require('./cipher/cipher.js');
var publicKeyEncrypter = require('./cipher/nacl/publicKeyEncrypter.js');
var publicKeyDecrypter = require('./cipher/nacl/publicKeyDecrypter.js');
var privateKeyEncrypter = require('./cipher/nacl/privateKeyEncrypter.js');
var privateKeyDecrypter = require('./cipher/nacl/privateKeyDecrypter.js');
var ecdh = require('./cipher/ecdh/ecdh.js');



Object.defineProperty(exports, 'CurveIds', {
	enumerable: true,
	get: function () { return keys.CurveIds; }
});
exports.IdED25519 = keys.IdED25519;
exports.IdNonSpecified = keys.IdNonSpecified;
exports.IdSECP256K1 = keys.IdSECP256K1;
exports.IdSECP256R1 = keys.IdSECP256R1;
exports.IdSR25519 = keys.IdSR25519;
exports.KindED25519 = keys.KindED25519;
exports.KindNoOp = keys.KindNoOp;
exports.KindSECP256K1 = keys.KindSECP256K1;
exports.KindSECP256R1 = keys.KindSECP256R1;
exports.KindSR25519 = keys.KindSR25519;
exports.isPublicKey = _public$2.isPublicKey;
exports.isPublicKeyEqual = _public$2.isPublicKeyEqual;
exports.secureRandom = rand.secureRandom;
exports.chainCodeFromDeriveIndex = hd$1.chainCodeFromDeriveIndex;
exports.ErrorUnsupportedKey = errors.ErrorUnsupportedKey;
exports.ED25519ExtendedPrivateKey = hd.ED25519ExtendedPrivateKey;
exports.ED25519PrivateKey = _private.ED25519PrivateKey;
exports.ED25519PublicKey = _public.ED25519PublicKey;
exports.deriveHardenedKey = derivation.ed25519DeriveHardenedKey;
exports.SECP256K1PrivateKey = _private$1.SECP256K1PrivateKey;
exports.SECP256K1PublicKey = _public$1.SECP256K1PublicKey;
exports.kindFromPublicKey = names.kindFromPublicKey;
exports.publicKeyFromKind = names.publicKeyFromKind;
exports.privateKeyFromBytes = bytes.privateKeyFromBytes;
exports.privateKeyToBytes = bytes.privateKeyToBytes;
exports.publicKeyFromBytes = bytes.publicKeyFromBytes;
exports.publicKeyToBytes = bytes.publicKeyToBytes;
exports.idFromPrivateKey = ids.idFromPrivateKey;
exports.idFromPublicKey = ids.idFromPublicKey;
exports.privateKeyFromId = ids.privateKeyFromId;
exports.publicKeyFromId = ids.publicKeyFromId;
exports.ED25519KeyExchange = ed25519.ED25519KeyExchange;
exports.KindNaClSecretKey = cipher.KindNaClSecretKey;
exports.NACLECDH = cipher.NACLECDH;
exports.NACLSK = cipher.NACLSK;
exports.PublicKeyEncrypter = publicKeyEncrypter.PublicKeyEncrypter;
exports.PublicKeyDecrypter = publicKeyDecrypter.PublicKeyDecrypter;
exports.PrivateKeyEncrypter = privateKeyEncrypter.PrivateKeyEncrypter;
exports.PrivateKeyDecrypter = privateKeyDecrypter.PrivateKeyDecrypter;
exports.fromPrivateKey = ecdh.fromPrivateKey;
exports.fromPublicKey = ecdh.fromPublicKey;
