// @mailchain/internal/browser/transport/payload/headers.js
import { headersMapFromBuffers, parseSignatureHeader, parseOriginHeader, createSignatureHeader, createOriginHeader } from '../serialization/headers.js';

const HEADER_CONTENT_ENCODING = 'Content-Encoding';
const HEADER_CONTENT_ENCRYPTION = 'Content-Encryption';
const HEADER_CONTENT_LENGTH = 'Content-Length';
const HEADER_CONTENT_LOCATION = 'Content-Location';
const HEADER_CONTENT_SIGNATURE = 'Content-Signature';
const HEADER_CONTENT_TYPE = 'Content-Type';
const HEADER_CREATED = 'Created';
const HEADER_ORIGIN = 'Origin';
class SerializableTransportPayloadHeaders {
    constructor(headers) {
        Object.defineProperty(this, "headers", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.headers = headers;
    }
    static FromEncryptedPayloadHeaders(headers) {
        return new this(headers);
    }
    static FromBuffer(buffer) {
        const { headers, invalidHeaders: _invalidHeaders } = headersMapFromBuffers(buffer, [
            HEADER_CONTENT_ENCODING,
            HEADER_CONTENT_ENCRYPTION,
            HEADER_CONTENT_LENGTH,
            HEADER_CONTENT_SIGNATURE,
            HEADER_CONTENT_TYPE,
            HEADER_CREATED,
            HEADER_ORIGIN,
        ]);
        return new this({
            ContentEncoding: headers.get(HEADER_CONTENT_ENCODING).toString(),
            ContentEncryption: headers.get(HEADER_CONTENT_ENCRYPTION).toString(),
            ContentLength: Number.parseInt(headers.get(HEADER_CONTENT_LENGTH).toString()),
            ContentSignature: parseSignatureHeader(headers.get(HEADER_CONTENT_SIGNATURE).toString()),
            ContentType: headers.get(HEADER_CONTENT_TYPE).toString(),
            Created: new Date(headers.get(HEADER_CREATED).toString()),
            Origin: parseOriginHeader(headers.get(HEADER_ORIGIN).toString()),
        });
    }
    ToBuffer() {
        const headers = [];
        headers.push(`${HEADER_CONTENT_ENCODING}: ${this.headers.ContentEncoding}`);
        headers.push(`${HEADER_CONTENT_ENCRYPTION}: ${this.headers.ContentEncryption}`);
        headers.push(`${HEADER_CONTENT_LENGTH}: ${this.headers.ContentLength}`);
        if (this.headers.ContentLocation) {
            headers.push(`${HEADER_CONTENT_LOCATION}: ${this.headers.ContentLocation}`);
        }
        headers.push(`${HEADER_CONTENT_SIGNATURE}: ${createSignatureHeader(this.headers.ContentSignature, this.headers.Origin)}`);
        headers.push(`${HEADER_CONTENT_TYPE}: ${this.headers.ContentType}`);
        headers.push(`${HEADER_CREATED}: ${this.headers.Created.toISOString()}`);
        headers.push(`${HEADER_ORIGIN}: ${createOriginHeader(this.headers.Origin)}`);
        return Buffer.from(headers.join('\r\n'), 'utf8');
    }
}

export { SerializableTransportPayloadHeaders };
