// @mailchain/encoding/node/index.js
'use strict';

var consts = require('./consts.js');
var ensure = require('./ensure.js');
var encoding = require('./encoding.js');
var hex = require('./hex.js');
var base32 = require('./base32.js');
var base58 = require('./base58.js');
var base58Check = require('./base58Check.js');
var base64 = require('./base64.js');
var utf8 = require('./utf8.js');



exports.EncodingTypes = consts.EncodingTypes;
exports.ensureDecoded = ensure.ensureDecoded;
exports.decode = encoding.decode;
exports.encode = encoding.encode;
exports.decodeHex = hex.decodeHex;
exports.decodeHexAny = hex.decodeHexAny;
exports.decodeHexZeroX = hex.decodeHexZeroX;
exports.encodeHex = hex.encodeHex;
exports.encodeHexZeroX = hex.encodeHexZeroX;
exports.isAnyHex = hex.isAnyHex;
exports.isHex = hex.isHex;
exports.isHexZeroX = hex.isHexZeroX;
exports.decodeBase32 = base32.decodeBase32;
exports.encodeBase32 = base32.encodeBase32;
exports.isBase32 = base32.isBase32;
exports.decodeBase58 = base58.decodeBase58;
exports.encodeBase58 = base58.encodeBase58;
exports.isBase58 = base58.isBase58;
exports.decodeBase58Check = base58Check.decodeBase58Check;
exports.encodeBase58Check = base58Check.encodeBase58Check;
exports.decodeBase64 = base64.decodeBase64;
exports.decodeBase64UrlSafe = base64.decodeBase64UrlSafe;
exports.encodeBase64 = base64.encodeBase64;
exports.encodeBase64UrlSafe = base64.encodeBase64UrlSafe;
exports.decodeUtf8 = utf8.decodeUtf8;
exports.encodeUtf8 = utf8.encodeUtf8;
